cloudinit/net/dhcp.py

```diff
@@ -251,13 +263,17 @@ def dhcp_discovery(dhclient_cmd_path, interface, cleandir):
             if ppid == 1:
                 LOG.debug('killing dhclient with pid=%s', pid)
                 os.kill(pid, signal.SIGKILL)
-                return parse_dhcp_lease_file(lease_file)
+                daemonized = True
+                break
         time.sleep(0.01)
 
-    LOG.error(
-        'dhclient(pid=%s, parentpid=%s) failed to daemonize after %s seconds',
-        pid_content, ppid, 0.01 * 1000
-    )
+    if not daemonized:
+        LOG.error(
+            'dhclient(pid=%s, parentpid=%s) failed to daemonize after %s '
+            'seconds', pid_content, ppid, 0.01 * 1000
+        )
+    if capture:
+        return parse_dhcp_lease_file(lease_file), err
     return parse_dhcp_lease_file(lease_file)
 
```

Comment:
I wonder if this would be easier if, instead of passing in a boolean and then having to store the return value somewhere to be accessed later, we made this an optional callable that will be called with the log output? We would still need to have the extra parameter through the stack, but we wouldn't need to change the way things are returned.

In my proposal, the new call signatures would look like this:

```py
def maybe_perform_dhcp_discovery(nic=None, dhcp_log_func=None):
def dhcp_discovery(dhclient_cmd_path, interface, cleandir, dhcp_log_func=False):
```

and this code would be simplified to:

```suggestion
    if dhcp_log_func is not None:
        dhcp_log_func(err)
    return parse_dhcp_lease_file(lease_file)
```

This would require adding an optional parameter to either `EphemeralDHCPv4` itself, or its `obtain_lease` method.

What do you think?

Reply:
That's a great idea, I'll go for it and commit my changes.